working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1165,6 +1165,71 @@ That is less important here since the only other case is the else branch.
 
 ## Static semantics
 
+### Refutable and irrefutable patterns and contexts
+
+Patterns appear inside a number of other constructs in the language. This
+proposal extends Dart to allow patterns in:
+
+* Local variable declarations.
+* For loop variable declarations.
+* Assignment expressions.
+* Switch statement cases.
+* A new switch expression form's cases.
+* A new pattern-if statement.
+
+When a pattern appears in a switch case, any variables bound by the pattern are
+only in scope in that case's body. If the pattern fails to match, the case body
+is skipped. This ensures that the variables can't be used when the pattern
+failed to match and they have no defined value. Likewise, the variables bound by
+a pattern-if statement's pattern are only in scope in the then branch. That
+branch is skipped if the pattern fails to match.
+
+When patterns appear in variable declarations and assignments, there is no
+natural control flow that can skip over uses of the variables if the pattern
+fails to match. For example:
+
+```dart
+main() {
+  var (a, b) = (1, 2);
+  print(a + b);
+}
+```
+
+What happens when `a` is printed in the example above?
+
+To avoid that, we restrict which patterns can be used in variable declarations
+and pattern assignments. Only *irrefutable* patterns that never fail to match
+are allowed in contexts where match failure can't be handled. For example, this
+is an error:
+
+```dart
+main() {
+  var (== 2, == 3) = (1, 2);
+}
+```
+
+We define an *irrefutable context* as the pattern in a
+`localVariableDeclaration`, `forLoopParts`, or `patternAssignment` or any of its
+subpatterns. It is a compile-time error if any of these patterns appear in an
+irrefutable context:
+
+*   Logical-or
```

Comment:
(Doesn't *have* to be irrefutable, but it's probably hard to do anything useful in an `or` that requires at least one branch to match. If the last case always works anyway, why not just use it for everything.)